### PR TITLE
chaos testing + memberlist clustering changes + govendor -> dep

### PR DESCRIPTION
**Pull request by Dieterbe**
needs a bit more work, mostly sharing for anyone who's curious what i'm up to.
see README.md and chaos_test.go for the highlevel overview

### CODE REVIEW

**Comment by Dieterbe** [comment]
one observed failure case is this:
```
=== RUN   TestClusterStartup
--- PASS: TestClusterStartup (24.94s)
=== RUN   TestClusterBaseIngestWorkload
--- PASS: TestClusterBaseIngestWorkload (14.40s)
=== RUN   TestQueryWorkload
--- PASS: TestQueryWorkload (60.01s)
=== RUN   TestIsolateOneInstance
--- FAIL: TestIsolateOneInstance (60.00s)
        chaos_test.go:143: expected mt4 to return either correct or erroring responses. got (chaos.checkResults) {
                 Mutex: (sync.Mutex) {
                  state: (int32) 0,
                  sema: (uint32) 0
                 },
                 valid: ([]int) (len=2 cap=2) {
                  (int) 664, // correct responses
                  (int) 532 // 503 responses
                 },
                 empty: (int) 0,
                 timeout: (int) 4,
                 other: (int) 0,
                 firstOther: (*chaos.response)(<nil>)
                }
```
test suite detected 4 timeouts, in mt4 logs i see 2 requests that would have timed out the client.

```
$ docker logs dockerchaos_metrictank4_1 2>&1 | grep -c 'Completed /render.*503 Service Unavailable in 5.[0-9]*s'
532
$ docker logs dockerchaos_metrictank4_1 2>&1 | grep Completed | rv 'µs' | rv 'ms' | rv 'Completed /render.*503 Service Unavailable in 5\.[0-9]*s' | rv '200 OK.*[0-4]\.[0-9]*s'
[Macaron] 2017-11-23 00:12:43: Completed /render?target=sum(some.id.of.a.metric.*)&format=json&from=-15s 200 OK in 5.017100176s
[Macaron] 2017-11-23 00:13:05: Completed /render?target=sum(some.id.of.a.metric.*)&format=json&from=-15s 503 Service Unavailable in 53.436061928s
[Macaron] 2017-11-23 00:13:05: Completed /render?target=sum(some.id.of.a.metric.*)&format=json&from=-15s 503 Service Unavailable in 53.186765272s

```

maybe the completed lines were not logged? but:
```
~/g/s/g/g/m/chaos ❯❯❯ docker logs dockerchaos_metrictank4_1 2>&1 | grep -c 'Started GET /render'
2200
~/g/s/g/g/m/chaos ❯❯❯ docker logs dockerchaos_metrictank4_1 2>&1 | grep -c 'Completed /render'
2200
```

**Comment by Dieterbe** [comment]
one thing I've just learned is with current approach where we cut off all traffic from mt4 to the others, the others will remove mt4, and mt4 keeps getting messages from the others, but refutes them, and while it fails instances after ping timeouts, it doesn't seem to remove them from its list. so after the partition recovers, mt4 just talks to them again and the cluster is healed.
I will need to update this to make the failure work both ways

**Comment by Dieterbe** [comment]
I would like to merge this PR, mainly because of these particular generally useful commits:
0bb76bc67edc784ea984d867da84ec3f5ffdf458
c80a8eda41c2d1d00d43e32a965a50d6e403f55e
7955734b6f06514a48e528ee97671496ccda32c8
7dd7c208cceff1c0a78401487da2055a1f746bbe

Note : 
* the vendor dir is currently a bit of a mess and `scripts/vendor_health.sh` complains and breaks the build. I will fix before merging (obviously)
* 2 commits have `DO NOT MERGE` in the message. these are only useful for chaos diagnosis and i'll remove those out of the branch that will ultimately be merged

File: api/middleware/tracer.go [hunk 1]
```diff
@@ -49,6 +51,13 @@ func Tracer(tracer opentracing.Tracer) macaron.Handler {
 		macCtx.MapTo(macCtx.Resp, (*http.ResponseWriter)(nil))
 
 		rw := macCtx.Resp.(*TracingResponseWriter)
+
+		headers := macCtx.Resp.Header()
+		spanStr := span.(*jaeger.Span).String()
```
Comment (woodsaj) [inline api/middleware/tracer.go]: You can get the traceID with
```
span.Context().(jaeger.SpanContext).TraceId().String()
```
this saves you having to encode the spanId to a  string, then split the traceId out from the span string.

**Comment by Dieterbe** [comment]
how does this look now? @woodsaj @replay . mergey mergey?

**Comment by Dieterbe** [comment]
@DanCech mentioned another issue: bind-address is needlessly validated when it's not set, or something like that

File: cluster/config.go [hunk 2]
```diff
@@ -17,47 +17,84 @@ var (
...
 
-	addr, err := net.ResolveTCPAddr("tcp", clusterBindAddr)
+	Mode = ModeType(mode)
+
+	if mode != ModeMulti {
+		return
+	}
+
+	var err error
+	swimBindAddr, err = net.ResolveTCPAddr("tcp", swimBindAddrStr)
```
Comment (DanCech) [inline cluster/config.go]: This is still being tested in every case though it's only used for `manual`, we should either only test it for manual config or use it for every config type

Comment (Dieterbe) [reply cluster/config.go]: fixed in b49c5e807dedb3fe0bb2b24050b191151986d737

**Pull request merged**